src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/AssemblyBuilder.cs

```diff
@@ -32,8 +33,59 @@ public ModuleBuilder DefineDynamicModule(string name)
+        /// </summary>
+        /// <param name="name">The name of the assembly.</param>
+        /// <param name="coreAssembly">The assembly that denotes the "system assembly" that houses the well-known types such as <see cref="object"/></param>
+        /// <param name="assemblyAttributes">A collection that contains the attributes of the assembly.</param>
+        /// <returns>An <see cref="AssemblyBuilder"/> that can be persisted.</returns>
+        /// <exception cref="ArgumentNullException">The <paramref name="name"/> or <paramref name="name.Name"/> or <paramref name="coreAssembly"/> is null.</exception>
+        /// <remarks>Currently the persisted assembly doesn't support running, need to save it and load back to run.</remarks>
+        public static AssemblyBuilder DefinePersistedAssembly(AssemblyName name, Assembly coreAssembly, IEnumerable<CustomAttributeBuilder>? assemblyAttributes = null)
+        {
+            ArgumentNullException.ThrowIfNull(name);
+            ArgumentNullException.ThrowIfNull(coreAssembly);
+
+            Type? assemblyType = Type.GetType("System.Reflection.Emit.AssemblyBuilderImpl, System.Reflection.Emit", throwOnError: true);
+
+            if (assemblyType == null)
+            {
+                throw new TypeLoadException(SR.Format(SR.TypeLoad_TypeNotFound, "System.Reflection.Emit.AssemblyBuilderImpl"));
+            }
+
+            ConstructorInfo con = assemblyType.GetConstructor(BindingFlags.NonPublic | BindingFlags.Instance, [typeof(AssemblyName), typeof(Assembly), typeof(IEnumerable<CustomAttributeBuilder>)])!;
+            return (AssemblyBuilder)con.Invoke([name, coreAssembly, assemblyAttributes]);
+        }
+
         protected abstract ModuleBuilder? GetDynamicModuleCore(string name);
 
+        /// <summary>
+        /// Serializes the assembly to stream.
+        /// </summary>
+        /// <param name="stream">The stream to which the assembly serialized.</param>
+        /// <exception cref="ArgumentNullException"> <paramref name="stream"/> is null </exception>
+        public void Save(Stream stream) => SaveCore(stream);
+
+        /// <summary>
+        /// Saves the assembly to disk.
+        /// </summary>
+        /// <param name="assemblyFileName">The file name of the assembly.</param>
+        /// <exception cref="ArgumentNullException"> <paramref name="assemblyFileName"/> is null </exception>
+        public void Save(string assemblyFileName)
+        {
+            ArgumentNullException.ThrowIfNull(assemblyFileName);
+
+            using var peStream = new FileStream(assemblyFileName, FileMode.Create, FileAccess.Write);
+            SaveCore(peStream);
+        }
+
+        /// <summary>
+        /// When implemented in derived type serializer the assembly to stream.
+        /// </summary>
+        /// <param name="stream">The stream to which the assembly serialized.</param>
+        protected abstract void SaveCore(Stream stream);
```

Comment:
Was it not possible for someone externally to derive from AssemblyBuilder in .NET 8 and earlier? Adding a new abstract method will break anyone who did.

Reply:
We made the `AssemblyBuilder` abstract in .NET 8, before it was sealed. So only .NET 8 can be considered, and I think it's unlikely somebody would have derived from `AssemblyBuilder` and build their own implementation.

Reply:
If it was possible to derive from AssemblyBuilder in .NET 8, then this is a binary breaking change. I think we need to make it virtual with the base implementation throwing.
cc: @terrajobst